### ISSUE TITLE
Slower performance in cp2k-8.1

### ISSUE THREAD

**Issue by JuntingYu1996** [cp2k/cp2k#1323]
I built the newest cp2k-8.1 and previous cp2k-7.1 on my cluster using intel-2018-update3 compiler (including icc, ifort, intel-mpi and MKL). Their porformances on a simple BOMD task are tested, and timing information are attached. It seems version 8.1 is slower than 7.1. I also noticed that in version 8.1, the grid part in src is rewritten in C, and the SUBROUTINEs **grid_integrate_task_list** and **grid_collocate_task_list** take up a lot of time, while in version 7.1 the SUBROUTINEs **integrate_v_rspace** and **calculate_rho_elec** are faster. 

Is there any suggestions that the performance of version 8.1 can be improved?

[8.1_timing.txt](https://github.com/cp2k/cp2k/files/5854234/8.1_timing.txt)

[7.1_timing.txt](https://github.com/cp2k/cp2k/files/5854233/7.1_timing.txt)

BTW, the arch files for compiling 8.1 and 7.1 are almost the same, except that in 8.1 arch file, the following line should be added accounting for grid part written in C.

CFLAGS += -std=c11 -qopenmp -I$(LIBXSMM)/include

**Comment by oschuett** [comment]
The collocate and integrate routines were indeed re-implemented in C before GPU support was added. Generally, we believe that the performance of the new implementation is [comparable](https://github.com/cp2k/cp2k/issues/1084#issuecomment-718779047) to the old one.

Maybe you can tell us a bit more about your setup, e.g. which basis set and how many OpenMP threads are you using?

**Comment by JuntingYu1996** [comment]
TZV2P-GTH basis sets with GTH pseudopotentials are used for all atoms. The xc functional is SCAN. 
In the SCF part, I used OT/IRAC algorithm with DIIS minimizer.
I used 112 MPI processes, with 1 OpenMP thread per MPI process for running cp2k.

**Comment by hfp** [comment]
> I used 112 MPI processes, with 1 OpenMP thread per MPI process for running cp2k.

Is this running MPI ranks on HT threads or is this an AP-platform?

**Comment by JuntingYu1996** [comment]
> > I used 112 MPI processes, with 1 OpenMP thread per MPI process for running cp2k.
> 
> Is this running MPI ranks on HT threads or is this an AP-platform?

Sorry, I do not have much knowledge about HT threads or AP-platform. The CPU model on my cluster is Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz. So I guess it might be hyper-threading?

**Comment by AdamsGo** [comment]
I compiled both the 7.1 and 8.1 version, with the same  Linux-x86-64-intel-minimal.psmp setting and the same Intel-parallel-studio/cluster.2018.4-intel-18.0.4 compiler on a supercomputer.
With one node (40 cores per node), 7.1 psmp completed the H20-64 benchmark test in 45.6 s, while 8.1 psmp completed this test in 66.0 s.
Can you share you arch file ?
[7.1.txt](https://github.com/cp2k/cp2k/files/5911447/7.1.txt)

[8.1.txt](https://github.com/cp2k/cp2k/files/5911449/8.1.txt)

**Comment by oschuett** [comment]
@AdamsGo, how many OpenMP threads are you using? I have a theory that [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L296) and [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L521) reduction are the culprits.

**Comment by alazzaro** [comment]
> I used 112 MPI processes, with 1 OpenMP thread per MPI process for running cp2k.

@oschuett , @JuntingYu1996 reported that he is using a single thread...

**Comment by AdamsGo** [comment]
> @AdamsGo, how many OpenMP threads are you using? I have a theory that [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L296) and [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L521) reduction are the culprits.
I have double checked, it's single thread.   I will try to compile 8.1 and 7.1 on my local machine to see  if it has to do
with the HPC parallel enviroment.

**Comment by AdamsGo** [comment]
> @AdamsGo, how many OpenMP threads are you using? I have a theory that [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L296) and [this](https://github.com/cp2k/cp2k/blob/master/src/grid/ref/grid_ref_task_list.c#L521) reduction are the culprits.

I have double checked, it's single thread. I will try to compile 8.1 and 7.1 on my local machine to see if it has to do
with the HPC parallel enviroment.

**Comment by alazzaro** [comment]
@AdamsGo We saw the same problem on other systems, no need to bother.

@oschuett we have to put in place a strategy to fix this performance regression, eventually we have to make a new release. Do you have a plan?

**Comment by oschuett** [comment]
> We saw the same problem on other systems, no need to bother.

I disagree because there are also a number of system where we didn't see it. For example on [Daint MC](https://dashboard.cp2k.org/archive/cray-xc50-daint-mc-psmp_perf/index_full.html) the H20-64 single thread benchmark did not slow down.

So, we should gather more data points until we understand what triggers it. If it's not the number of threads, maybe it's the compiler?

**Comment by alazzaro** [comment]
The example you are referring to is probably too small (only 27s) to see a real effect. I don't have an old enough output to compare with, but I don't think we need to debug where the problem is small, so this test doesn't conclude anything...
The slowdown is much more evident for the H2O-32 RI-MP2 / Ri-RPA. 

So the problem exists, as reported by @AdamsGo and @marci73  too, and that's enough for me to start to worry and investigate.

I did a comparison of the 7.1 and 8.1 outputs provided by @AdamsGo , these are the CP2K timers (2nd column 7.1, 3rd is 8.1):

```
cp_fm_syevd_base                   6.348     ----------
calculate_rho_elec                 6.499     0.027
integrate_v_rspace                 8.171     0.002
CP2K_Total                        45.678    66.006

Remaining entries in  ==> 8.1.txt
multiply_cannon_loop               0.169
cp_fm_diag_elpa_base               4.704
grid_collocate_task_list           9.492
grid_integrate_task_list           9.735
cp_fm_gemm                        21.449
```

My understanding is that `calculate_rho_elec` and `integrate_v_rspace` are basically "replaced" (I mean the execution time) by the `grid_*` functions (few seconds slower, still reasonable). Then I assume that 7.1 doesn't use ELPA (it has the syevd call), while 8.1 does (please @AdamsGo confirm). The big difference in performance is cp_fm_gemm, which is only in the 8.1 call. Is this due to the `grid` changes? What kind of gemm is this, something we can replace with a COSMA call?

**Comment by oschuett** [comment]
> So the problem exists...

I never denied that. All I'm saying is that we first need to understand the problem before we can fix it.

> The example you are referring to is probably too small (only 27s) to see a real effect.

No, a significant performance regression in collocate / integrate should still show up. These two operations take up about 50% of the runtime and the error bars are quite tight.

> The slowdown is much more evident for the H2O-32 RI-MP2 / Ri-RPA.

Yes, there is something special going on with those RI benchmarks and I certainly intent to find out. Nevertheless, they only slowed down by 20%, which is much less than the 45% reported by @AdamsGo.

> The big difference in performance is cp_fm_gemm, which is only in the 8.1 call. Is this due to the grid changes?

Good catch! This has definitely nothing to do with my grid changes. Yes, COSMA would have been my first suspect too (#1303).

**Comment by AdamsGo** [comment]
@alazzaro To clarify, I just recomplied 7.1, 8.1 and the developing version with:

>             module load intel-mkl/2020.1.217-intel-19.1.1 
>             module load openmpi/3.1.5-gcc-9.2.0 
>             module load gcc/9.2.0-gcc-4.8.5 
>            ./install_cp2k_toolchain.sh  --with-mkl=system  --with-openmpi=system --with-scalapack=no  --with-ptscotch=no  --with-superlu=no  --with-pexsi=no --with-quip=no --with-plumed=install --with-spfft=no --with-sirius=no --with-libint=no --with-libxc=install --with-elpa=no --with-fftw=no  --with-libvdwxc=no
>  
the with-libxc=install option was only applied for the developing version, without it the compliation would fail. 

I  performed benchmark test with the H2O-128.inp setting with export OMP_NUM_THREADS=1.

With a single node (40 cores), 7.1  fininshed the test with 113.119 s,  8.1 finished the test with 148.89 s
and  the developing version took 195 s.  
I am new to the linux system and cp2k, so I might made mistakes and as @oschuett pointed out  
on Daint MC the H20-64 single thread benchmark did not slow down. I would not be surprised if it's
related to my HPC settings.


**UPDATE: with the --with-cosma=no flag, the 8.1 version finished the H2O-128 test in 124s and 130s in two sepearated tests.**



The attached timing data has been sorted according to the average self time.
[git_current_mkl.txt](https://github.com/cp2k/cp2k/files/5916414/git_current_mkl.txt)
[8.1_mkl.txt](https://github.com/cp2k/cp2k/files/5916415/8.1_mkl.txt)
[7.1_mkl.txt](https://github.com/cp2k/cp2k/files/5916413/7.1_mkl.txt)

**Comment by alazzaro** [comment]
@AdamsGo I assume this is a different test with respect to the previous outputs you provided... Let's stick with it and try to understand the performance problem.

So, as I understand from your UPDATE, the problem is COSMA. As @oschuett pointed out, this is a known issue (#1303 ). I do suggest to use `--with-cosma=no`. 

At this point, I suggest to use ELPA (it seems it gives some speedup) and libxsmm. 
Could you rerun the 3 cases and provided the full (standard) CP2K output? I have a script to make the comparison...

**Comment by AdamsGo** [comment]
Ok, sticking to my previous comment (H2O-128 benchmark test).

[7.1_log.txt](https://github.com/cp2k/cp2k/files/5916928/7.1_log.txt)
[8.1_no_cosma_log.txt](https://github.com/cp2k/cp2k/files/5916927/8.1_no_cosma_log.txt)
[8.1_with_cosma_log.txt](https://github.com/cp2k/cp2k/files/5916930/8.1_with_cosma_log.txt)
[git_log_40_cpu.txt](https://github.com/cp2k/cp2k/files/5916926/git_log_40_cpu.txt)

**Comment by alazzaro** [comment]
Do you mind recompiling the git master version without COSMA too?

**Comment by AdamsGo** [comment]
Here it is, almost the same performance as 8.1 without cosma.
[git_no_cosma.txt](https://github.com/cp2k/cp2k/files/5917238/git_no_cosma.txt)

**Comment by alazzaro** [comment]
OK, digging in the new outputs, I see that with COSMA git_log goes up in the following timers
(order of values: 7.1, 8.1 w/o COSMA, 8.1 w/ COSMA, git_log w/ COSMA):

```
cp_fm_gemm                         0.809     0.755     1.547     14.432
cp_fm_cholesky_invert              1.810     1.712     23.432    58.106
```

This is clearly COSMA (even in Cholesky). So we understood COSMA is not good to have it here. I wonder what will happen if you change the number of threads, but this is another discussion...

Then, let's consider 7.1 versus 8.1_no_cosma, these are the hotspots:

```
dbcsr_make_untransposed_blocks     0.127     1.885
integrate_v_rspace                15.957     0.004
calculate_rho_elec                18.769     0.064
CP2K_Total                       113.119   124.52

Remaining entries in  ==> 8.1_no_cosma_log.txt
grid_collocate_task_list          25.737
grid_integrate_task_list          26.583

```

Overall, CP2K 8.1 is faster in several timers, except the ones I'm reporting above.
So, I would say that the 10 seconds in the CP2K_Total are entirely due to the new grid_ stuff (minus the integrate_v_rspace and calculate_rho_elec).

**Comment by hfp** [comment]
Indeed, given "7.1 finished the test with 113.119 s" and "8.1 version finished the H2O-128 test in 124..130s" (no COSMA), there are about 11..17 s missing like 10-15% performance regression due to "new grid_ stuff (minus the integrate_v_rspace and calculate_rho_elec)". Unfortunately, I believe this is still relevant beside of what's recovered when omitting COSMA.

**Comment by alazzaro** [comment]
> Indeed, given "7.1 finished the test with 113.119 s" and "8.1 version finished the H2O-128 test in 124..130s" (no COSMA), there are about 11..17 s missing like 10-15% performance regression due to "new grid_ stuff (minus the integrate_v_rspace and calculate_rho_elec)". Unfortunately, I believe this is still relevant beside of what's recovered when omitting COSMA.

I agree, this is A LOT relevant for me (i.e. LUMI-C, Archer2, Daint-MC...)

**Comment by hfp** [comment]
Yes, even recent PRACE CP2K benchmark is about H2O-512 if I am not mistaken.

**Comment by alazzaro** [comment]
> Here it is, almost the same performance as 8.1 without cosma.
> [git_no_cosma.txt](https://github.com/cp2k/cp2k/files/5917238/git_no_cosma.txt)

I cannot spot where the 5s difference comes from, besides some minor changes. The most significant is:

init_cphi_and_sphi

which is slower (by 1s) in git_master version. Overall, I would say 8.1 and git_master are compatible.

**Comment by oschuett** [comment]
> 10-15% performance regression due to "new grid_ stuff

Yes, that's more in line with my own observations. I'm certainly committed to optimizing the new grid code and I'm pretty optimistic that we'll eventually outperform 7.1. However, at this point it's mostly a research effort and I can't make any promises. We should therefore accept the 8.1 performance as our new base line.

> This is clearly COSMA (even in Cholesky). So we understood COSMA is not good to have it here.

It's very unfortunate that we didn't catch this with our performance tests. They should really always use the latest toolchain.

**Comment by mtaillefumier** [comment]
> OK, digging in the new outputs, I see that with COSMA git_log goes up in the following timers
> (order of values: 7.1, 8.1 w/o COSMA, 8.1 w/ COSMA, git_log w/ COSMA):
> 
> ```
> cp_fm_gemm                         0.809     0.755     1.547     14.432
> cp_fm_cholesky_invert              1.810     1.712     23.432    58.106
> ```
> 
> This is clearly COSMA (even in Cholesky). So we understood COSMA is not good to have it here. I wonder what will happen if you change the number of threads, but this is another discussion...
> 
> Then, let's consider 7.1 versus 8.1_no_cosma, these are the hotspots:
> 
> ```
> dbcsr_make_untransposed_blocks     0.127     1.885
> integrate_v_rspace                15.957     0.004
> calculate_rho_elec                18.769     0.064
> CP2K_Total                       113.119   124.52
> 
> Remaining entries in  ==> 8.1_no_cosma_log.txt
> grid_collocate_task_list          25.737
> grid_integrate_task_list          26.583
> ```
> 
> Overall, CP2K 8.1 is faster in several timers, except the ones I'm reporting above.
> So, I would say that the 10 seconds in the CP2K_Total are entirely due to the new grid_ stuff (minus the integrate_v_rspace and calculate_rho_elec).

@alazzaro could you try to add this in the global section (if you run on cpu)

&GRID
BACKEND cpu
APPLY_CUTOFF false
&END

and rerun the test. Or I can do it if I have the input.

**Comment by alazzaro** [comment]
@mtaillefumier It is not me who run the test... I'm only comparing the inputs @AdamsGo provided to me.
This is a question for @AdamsGo 
However, I think he is not using GPUs...

**Comment by mtaillefumier** [comment]
> @mtaillefumier It is not me who run the test... I'm only comparing the inputs @AdamsGo provided to me.
> This is a question for @AdamsGo
> However, I think he is not using GPUs...

@alazzaro Sorry for the mistake. I think @AdamsGo  should try this on cpu only machines. The collocate/integrate functions will use dgemm for the computations

**Comment by alazzaro** [comment]
> > @mtaillefumier It is not me who run the test... I'm only comparing the inputs @AdamsGo provided to me.
> > This is a question for @AdamsGo
> > However, I think he is not using GPUs...
> 
> @alazzaro Sorry for the mistake. I think @AdamsGo should try this on cpu only machines. The collocate/integrate functions will use dgemm for the computations

You are right! I misread GPU instead of CPU ;)

**Comment by alas-go** [comment]
I think it is the the library for low scaling method, which contains 3 library. one of them is the superlu. it would not been installed by default and the package of superlu-6.1.0 or 6.3.0 could not been compiled due to the omp / ompi settings both for openmpi and intelmpi.  I report the problem but no progress has been made.

**Comment by alazzaro** [comment]
> I think it is the the library for low scaling method, which contains 3 library. one of them is the superlu. it would not been installed by default and the package of superlu-6.1.0 or 6.3.0 could not been compiled due to the omp / ompi settings both for openmpi and intelmpi. I report the problem but no progress has been made.

@alas-go I'm not sure if your comment is meant for this issue... Could you confirm?

**Comment by oschuett** [comment]
This was fixed via #1398. The orthorhombic collocate/integrate on CPU is now 5-10% faster compared to cp2k 7.1.
